commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/Norms.java

```diff
@@ -0,0 +1,427 @@
+        return Summation.value(
+                Math.abs(x),
+                Math.abs(y),
+                Math.abs(z));
+    }
+
+    /** Compute the Manhattan norm (also known as the Taxicab norm or L1 norm) of the given values.
+     * The result is equal to \(|v_0| + ... + |v_i|\), i.e., the sum of the absolute values of the input elements.
+     *
+     * <p>Special cases:
+     * <ul>
+     *  <li>If any value is NaN, then the result is NaN.</li>
+     *  <li>If any value is infinite and no value is NaN, then the result is positive infinity.</li>
+     *  <li>If the array is empty, then the result is 0.</li>
+     * </ul>
+     * @param v input values
+     * @return Manhattan norm
+     * @see <a href="https://en.wikipedia.org/wiki/Norm_(mathematics)#Taxicab_norm_or_Manhattan_norm">Manhattan norm</a>
+     */
+    public static double manhattan(final double[] v) {
+        double sum = 0d;
+        double comp = 0d;
+
+        for (int i = 0; i < v.length; ++i) {
+            final double x = Math.abs(v[i]);
+            final double sx = sum + x;
+            comp += ExtendedPrecision.twoSumLow(sum, x, sx);
+            sum = sx;
+        }
+
+        return Summation.summationResult(sum, comp);
+    }
+
+    /** Compute the Euclidean norm (also known as the L2 norm) of the arguments. The result is equal to
+     * \(\sqrt{x^2 + y^2}\). This method correctly handles the possibility of overflow or underflow
+     * during the computation.
+     *
+     * <p>Special cases:
+     * <ul>
+     *  <li>If either value is NaN, then the result is NaN.</li>
+     *  <li>If either value is infinite and the other value is not NaN, then the result is positive infinity.</li>
+     * </ul>
+     *
+     * <p><strong>Comparison with Math.hypot()</strong>
+     * <p>While not guaranteed to return the same result, this method does provide similar error bounds to
+     * the JDK's Math.hypot() method and may run faster on some JVMs.
+     * @param x first input
+     * @param y second input
+     * @return Euclidean norm
+     * @see <a href="https://en.wikipedia.org/wiki/Norm_(mathematics)#Euclidean_norm">Euclidean norm</a>
+     */
+    public static double euclidean(final double x, final double y) {
+        final double xabs = Math.abs(x);
+        final double yabs = Math.abs(y);
+
+        final double max = Math.max(xabs, yabs);
+
+        // if the max is not finite, then one of the inputs must not have
+        // been finite
+        if (!Double.isFinite(max)) {
+            if (Double.isNaN(x) || Double.isNaN(y)) {
+                return Double.NaN;
+            }
+            return Double.POSITIVE_INFINITY;
+        }
+
+        // compute the scale and rescale values
+        final double scale;
+        final double rescale;
+        if (max > LARGE_THRESH) {
+            scale = SCALE_DOWN;
+            rescale = SCALE_UP;
+        } else if (max < SAFE_SCALE_UP_THRESH) {
```

Comment:
@aherbert, this is slightly different than the implementation in the ExtendedPrecisionLinearCombination algorithm in examples-jmh. In that class, we scale up if the max is under a threshold for safely scaling up _and_ the min value is less than the threshold for very small numbers. My thought here is that we only need to check the max. If the max value is small enough that we can scale up without overflow, then we may as well just go ahead and do it and not bother checking (or even computing) the min. Let me know if you agree with this assessment.

Reply:
The logic makes sense since you only have two numbers. Note you can change this to a new value:
```java
SAFE_SCALE_UP_THRESH = 0x1.0p-100;
```
You are scaling by 2^600. So this value scaled would be 2^500 and safe against overflow when squared.

With two numbers you do not need to do the computation if the difference between them is large because max^2 + min^2 is mute:
```java
if (max - min > 0x1.0p54) {
    return max;
}
```
So knowing min is useful for a fast exit.

I'll do a full review of this later. But the thresholds should be changed to 2^496 from 2^500 for the array case.

Reply:
Why the change for the array case? Also, should/could that apply to the other cases as well?

Reply:
Because if you have a lot of memory you can pass in an array of 2^31-1 doubles. If these all have a value of Double.MAX_VALUE then you will overflow during the sum unless you lower the threshold. When you only have 2 or 3 doubles then you could make the threshold higher.

I think you need 2^496 for the array case and 2^510 for the case of 3 inputs. This would be 2^992 or 2^1020 when squared. You can add respectively 2^31 of the former and 3 of the latter.

Reply:
I should add that due to the extended precision summation you do not want it to sum to infinite. This will lead to errors in the round-off computation that creates NaN. So even if the final result would be infinite you do not want it to return NaN. So the thresholds should allow the summation to complete to a finite value. The final scaling can then create the infinite if the result is not possible within the range of a double.

Reply:
I've updated the threshold and added the early exit check for Euclidean 2D. I don't see a compelling reason to use two separate thresholds, so all Euclidean methods use the 2^496 threshold. The early exit in the 2D case gave a small improvement in performance in the JMH benchmark, small enough where the outliers of the before and after case overlap.

Reply:
Let me know if you see any other changes that should be made.

Reply:
@aherbert, did you still want to review this or do you think it's ready to go?

Reply:
Sorry, I have been busy. I will try and find time this weekend.

Reply:
No worries! I was just making sure there wasn't something I was missing here.